lab-nassir/test/cafe-route-test.js

```diff
@@ -0,0 +1,385 @@
+'use strict';
+
+const PORT = process.env.PORT || 3000;
+process.env.MONGODB_URI = 'mongodb://localhost/cattest';
+
+const expect = require('chai').expect;
+const request = require('superagent');
+const Cafe = require('../model/cafe');
+
+require('../server.js');
+
+const url = `http://localhost:${PORT}`;
+
+const exampleCafe = {
+  name: 'Test Cafe',
+  address: 'Test Address',
+  timestamp: new Date(),
+};
+
+const exampleCafeTwo = {
+  name: 'Test Cafe Two',
+  address: 'Test Address Two',
+};
+
+describe('Testing routes on /api/cafe', function(){
+
+  describe('Testing POST /api/cafe', function(){
+
+    after(done => {
+      if(this.tempCafe) {
+        Cafe.remove({})
+        .then(() => done())
+        .catch(done);
+        return;
+      }
+      done();
+    });
+
+    describe('Testing POST /api/cafe with VALID BODY', function(){
+      it('Should return a status code of 200 and a valid cat cafe', done => {
+        request.post(`${url}/api/cafe`)
+        .send(exampleCafe)
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(200);
+          expect(res.body.name).to.equal('Test Cafe');
+          expect(res.body.address).to.equal('Test Address');
+          this.tempCafe = res.body;
+          done();
+        });
+      });
+    });
+
+    describe('Testing POST /api/cafe with NO BODY or INVALID BODY', function(){
+      it('Should return a status code of 400 and an error message', done => {
+        request.post(`${url}/api/cafe`)
+        .set('Content-type', 'application/json')
+        .send('{')
+        .end((err, res) => {
+          expect(res.status).to.equal(400);
+          expect(res.text).to.equal('SyntaxError');
+          done();
+        });
+      });
+    });
+  });
+
+  describe('Testing GET /api/cafe', function(){
+
+    describe('Testing GET /api/cafe with VALID ID', function(){
+
+      before(done => {
+        new Cafe(exampleCafe).save()
+        .then(cafe => {
+          this.tempCafe = cafe;
+          done();
+        })
+        .catch(done);
+        return;
+      });
+
+      after(done => {
+        if(this.tempCafe) {
+          Cafe.remove({})
+          .then(() => done())
+          .catch(done);
+          return;
+        }
+        done();
+      });
+
+      it('Should return a status code of 200 and a cat cafe', done => {
+        request.get(`${url}/api/cafe/${this.tempCafe._id}`)
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(200);
+          expect(res.body.name).to.equal('Test Cafe');
+          expect(res.body.address).to.equal('Test Address');
+          done();
+        });
+      });
+    });
+
+    describe('Testing GET /api/cafe with INVALID ID', function(){
+
+      before(done => {
```

Comment:
You could set up your test environment for all GET tests in the outer describe block and thus not have to repeat yourself.